### PR TITLE
core: Set default service account on all Ceph daemons

### PR DESCRIPTION
**Pull request by parth-gr**
When a private docker registry is used and an image pull secret is specified in the chart, the pods with default Service Account fail to pull the image due to authentication issues. Added rook-ceph-default service account and modify the pods specifications by adding the serviceAccountName.

Closes: https://github.com/rook/rook/issues/6673

(cherry picked from commit 737fb099feafa01489b233cb64f889c61b3b6016)

**Which issue is resolved by this Pull Request:**
Resolves # https://github.com/rook/rook/issues/12786

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

File: Documentation/Getting-Started/Prerequisites/authenticated-registry.md
Comment (BlaineEXE) [inline Documentation/Getting-Started/Prerequisites/authenticated-registry.md:50]: Typo I think

```suggestion
  name: rook-ceph-default
```

File: pkg/operator/ceph/cluster/rbd/spec_test.go
Comment (BlaineEXE) [inline pkg/operator/ceph/cluster/rbd/spec_test.go:23]: Unnecessary newline

**Review by BlaineEXE — COMMENTED** [review]
I think this looks good overall. The only conern I have is for daemons that aren't often used that we might miss in CI tests.

I see the crash collector, rbdmirror, and cephfs mirror daemons here, which is great. The cmd-reporter has its own SA already.

What about the metrics collector or discovery daemon? Do those need updated, or do they have a specific SA? Are there other daemons that might be easy to miss?

**Comment by parth-gr** [comment]
@BlaineEXE does detect jb also need this service account https://github.com/rook/rook/blob/master/pkg/operator/k8sutil/cmdreporter/cmdreporter.go#L298, there is a cyclic dependency of package import to , so we would to manually add the name

File: pkg/operator/ceph/cluster/nodedaemon/pruner.go
Comment (parth-gr) [inline pkg/operator/ceph/cluster/nodedaemon/pruner.go:114]: and also is this fine to add with this cron job

**Review by BlaineEXE — APPROVED** [review]
I think this looks good. I have mixed thoughts about backporting to 1.13, especially since it could cause breakages for some users, and we might not have a fix for many weeks.

**Review by travisn — CHANGES_REQUESTED** [review]
> I think this looks good. I have mixed thoughts about backporting to 1.13, especially since it could cause breakages for some users, and we might not have a fix for many weeks.

Good question on the timing. I would propose that we wait to merge this to master until after the new year, to avoid the risk of regression over the holidays. After the new year, let's finalize merging it and whether there is risk to back to 1.13, or if it should wait for 1.14. I'm leaning towards a 1.13 backport, but want to let it settle in master first for a bit. Adding changes requested to hold it for now...

**Comment by parth-gr** [comment]
> > I think this looks good. I have mixed thoughts about backporting to 1.13, especially since it could cause breakages for some users, and we might not have a fix for many weeks.
> 
> Good question on the timing. I would propose that we wait to merge this to master until after the new year, to avoid the risk of regression over the holidays. After the new year, let's finalize merging it and whether there is risk to back to 1.13, or if it should wait for 1.14. I'm leaning towards a 1.13 backport, but want to let it settle in master first for a bit. Adding changes requested to hold it for now...

Sure no problem, as per time permit we can have it in.

**Comment by BlaineEXE** [comment]
> @BlaineEXE does detect job also need this service account https://github.com/rook/rook/blob/master/pkg/operator/k8sutil/cmdreporter/cmdreporter.go#L298, there is a cyclic dependency of package import to , so we would to manually add the name

No. The command reporter already has it's own SA, though it is optional to use:
- https://github.com/rook/rook/blob/0a538bfc37f1c6e5e5dfa6936d9297b37553201a/pkg/operator/ceph/controller/network.go#L261
- https://github.com/rook/rook/blob/934aa910566c0176945ebf819d1ce24b5295bb4f/pkg/operator/ceph/controller/version.go#L83

If there are other uses of the cmdreporter, it would be good to make sure the SA is used.

File: Documentation/Getting-Started/Prerequisites/authenticated-registry.md
Comment (travisn) [inline Documentation/Getting-Started/Prerequisites/authenticated-registry.md:6]: Let's remove the sentence that mentions the specific service accounts here, and just refer them to the next section.
```suggestion
add an `imagePullSecret` to all relevant service accounts. See the next section for the required service accounts.
```

**Review by travisn — CHANGES_REQUESTED** [review]
Two final suggestions:
1. Rebase to make sure it's still passing the latest CI
2. Please test it manually on Openshift since the CI does not cover it. A basic creation of the cluster, filesystem, and object store should be sufficient to verify the new default service account

**Comment by travisn** [comment]
Let's wait on v1.14 for this update, it seems too fundamental of a change for a backport, even though not complex.

File: deploy/examples/odf-cli
Comment (travisn) [inline deploy/examples/odf-cli:1]: Please revert this `odf-cli` submodule, I assume it was not intended.

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/operator/k8sutil/serviceaccount.go
Comment (subhamkrai) [inline pkg/operator/k8sutil/serviceaccount.go:2]:
```suggestion
Copyright 2024 The Rook Authors. All rights reserved.
```
😄

**Review by subhamkrai — CHANGES_REQUESTED** [review]

File: deploy/examples/odf-cli
Comment (parth-gr) [reply deploy/examples/odf-cli:1]: Yaa by mistake, fixed thanks

File: pkg/operator/k8sutil/serviceaccount.go
Comment (parth-gr) [reply pkg/operator/k8sutil/serviceaccount.go]: Updated:)

File: Documentation/Getting-Started/Prerequisites/authenticated-registry.md
Comment (parth-gr) [inline Documentation/Getting-Started/Prerequisites/authenticated-registry.md:35]: I don't see the option in operator or cluster.yaml to update the service account do we have any suck options? @travisn

File: deploy/examples/common.yaml
Comment (travisn) [inline deploy/examples/common.yaml:1166]: No need for the `chart` label in common.yaml
```suggestion
```

**Review by travisn — CHANGES_REQUESTED** [review]

File: deploy/examples/common.yaml
Comment (parth-gr) [reply deploy/examples/common.yaml:1166]: Make sense updated

File: pkg/operator/k8sutil/serviceaccount.go
Comment (travisn) [inline pkg/operator/k8sutil/serviceaccount.go:20]: nit: How about defining this const in k8sutil.go? IMO it's not usually necessary to create a new go file if there is somewhere else to include something small like this.

**Review by travisn — APPROVED** [review]

File: pkg/operator/k8sutil/serviceaccount.go
Comment (parth-gr) [reply pkg/operator/k8sutil/serviceaccount.go]: Updated

**Review by subhamkrai — APPROVED** [review]

File: deploy/examples/common.yaml
Comment (travisn) [reply deploy/examples/common.yaml:1166]: Still need to push the change?

Comment (parth-gr) [reply deploy/examples/common.yaml]: Finally updated

**Pull request merged**
